File: python/src/benchmark/data_v2/generator.py
"""make_one generators for Data Model v2 types."""

from __future__ import annotations

from typing import Any

from .models import (
    Document,
    DocumentItem,
    DocumentMeta,
    Event,
    EventAttr,
    Message,
    Strings,
    Telemetry,
)
from .prng import XorShift64, mix_seed

# Fixed epoch base so runs are stable (not wall clock).
_BASE_TS_MS = 1_704_067_200_000  # 2024-01-01T00:00:00Z


def make_one(
    type_id: str,
    type_config: dict[str, Any],
    seed: int,
    instance_index: int = 0,
) -> Any:
    """Build one instance. type_config must already be resolved (no all_available)."""
    rng = XorShift64(mix_seed(seed, type_id, instance_index))
    if type_id == "message":
        return _make_message(rng, type_config)
    if type_id == "document":
        return _make_document(rng, type_config)
    if type_id == "telemetry":
        return _make_telemetry(rng, type_config)
    if type_id == "strings":
        return _make_strings(rng, type_config)
    if type_id == "event":
        return _make_event(rng, type_config)
    raise ValueError(f"unknown type_id: {type_id!r}")


def instances_for_cell(
    type_id: str,
    type_config: dict[str, Any],
    seed: int,
    data_type_instance_count: int,
) -> list[Any]:
    return [
        make_one(type_id, type_config, seed, i)
        for i in range(data_type_instance_count)
    ]


def _slen(cfg: dict[str, Any]) -> tuple[int, int]:
    sl = cfg.get("string_len") or {}
    return int(sl.get("min", 3)), int(sl.get("max", 16))


def _irange(cfg: dict[str, Any]) -> tuple[int, int]:
    ir = cfg.get("int_range") or {}
    return int(ir.get("min", 0)), int(ir.get("max", 1_000_000))


def _make_message(rng: XorShift64, cfg: dict[str, Any]) -> Message:
    lo, hi = _irange(cfg)
    smin, smax = _slen(cfg)
    # field_count reserved for future variable width; default maps to fixed Message
    return Message(
        f_bool=rng.next_bool(),
        f_int32=rng.next_int(lo, min(hi, 2**31 - 1)),
        f_int64=rng.next_int(lo, hi),
        f_float64=rng.next_f64() * 1000.0,
        f_string=rng.word(smin, smax),
        f_bool_2=rng.next_bool(),
        f_int32_2=rng.next_int(lo, min(hi, 2**31 - 1)),
        f_string_2=rng.word(smin, smax),
    )


def _make_document(rng: XorShift64, cfg: dict[str, Any]) -> Document:
    children = int(cfg.get("children", 8))
    smin, smax = _slen(cfg)
    items = [
        DocumentItem(
            sku=rng.word(smin, smax),
            qty=rng.next_int(1, 100),
            price_minor=rng.next_int(0, 100_000),
        )
        for _ in range(children)
    ]
    return Document(
        id=rng.word(8, 12),
        status=rng.next_int(0, 5),
        meta=DocumentMeta(region=rng.word(2, 4), version=rng.next_int(1, 10)),
        items=items,
    )


def _make_telemetry(rng: XorShift64, cfg: dict[str, Any]) -> Telemetry:
    points = int(cfg.get("points", 32))
    tag_count = int(cfg.get("tag_count", 2))
    smin, smax = _slen(cfg)
    tags = [rng.word(smin, smax) for _ in range(tag_count)]
    number_type = cfg.get("number_type", "float64")
    if number_type == "int64":
        values = [float(rng.next_int(0, 10_000)) for _ in range(points)]
    else:
        values = [rng.next_f64() * 100.0 for _ in range(points)]
    return Telemetry(
        source=rng.word(smin, smax),
        ts=_BASE_TS_MS + rng.next_int(0, 86_400_000),
        tags=tags,
        values=values,
    )


def _make_strings(rng: XorShift64, cfg: dict[str, Any]) -> Strings:
    count = int(cfg.get("count", 32))
    smin, smax = _slen(cfg)
    dup = float(cfg.get("duplication", 0.0))
    pool: list[str] = []
    items: list[str] = []
    for _ in range(count):
        if pool and rng.next_f64() < dup:
            items.append(pool[rng.next_int(0, len(pool) - 1)])
        else:
            w = rng.word(smin, smax)
            pool.append(w)
            items.append(w)
    return Strings(items=items)


def _make_event(rng: XorShift64, cfg: dict[str, Any]) -> Event:
    attr_count = int(cfg.get("attr_count", 4))
    smin, smax = _slen(cfg)
    attrs = [
        EventAttr(key=rng.word(smin, smax), value=rng.word(smin, smax))
        for _ in range(attr_count)
    ]
    return Event(
        event_id=rng.word(8, 12),
        event_type=rng.word(smin, smax),
        occurred_at=_BASE_TS_MS + rng.next_int(0, 86_400_000),
        producer=rng.word(smin, smax),
        attrs=attrs,
    )

File: c-sharp/src/TestData/V2/Generator.cs
// make_one for Data Model v2 — aligned with python data_v2/generator.py
using System;
using System.Collections.Generic;
using System.Text.Json;

namespace GLD.SerializerBenchmark.TestData.V2
{
    public static class Generator
    {
        const long BaseTsMs = 1_704_067_200_000L; // 2024-01-01T00:00:00Z

        public static object MakeOne(string typeId, JsonElement typeConfig, int seed, int instanceIndex = 0)
        {
            var rng = new Rng(MixSeed((ulong)seed, typeId, instanceIndex));
            return typeId switch
            {
                "message" => MakeMessage(rng, typeConfig),
                "document" => MakeDocument(rng, typeConfig),
                "telemetry" => MakeTelemetry(rng, typeConfig),
                "strings" => MakeStrings(rng, typeConfig),
                "event" => MakeEvent(rng, typeConfig),
                _ => throw new ArgumentException($"unknown type_id: {typeId}")
            };
        }

        public static (object data, Type dataType, List<Type> secondary) BuildPayload(
            string typeId, JsonElement typeConfig, int n, int seed)
        {
            if (n < 1) n = 1;
            switch (typeId)
            {
                case "message":
                {
                    var list = new List<Message>(n);
                    for (int i = 0; i < n; i++) list.Add((Message)MakeOne(typeId, typeConfig, seed, i));
                    if (n == 1) return (list[0], typeof(Message), new List<Type>());
                    return (new BatchMessage { Items = list }, typeof(BatchMessage), new List<Type> { typeof(Message) });
                }
                case "event":
                {
                    var list = new List<Event>(n);
                    for (int i = 0; i < n; i++) list.Add((Event)MakeOne(typeId, typeConfig, seed, i));
                    if (n == 1) return (list[0], typeof(Event), new List<Type> { typeof(EventAttr) });
                    return (new BatchEvent { Items = list }, typeof(BatchEvent),
                        new List<Type> { typeof(Event), typeof(EventAttr) });
                }
                case "document":
                {
                    var list = new List<Document>(n);
                    for (int i = 0; i < n; i++) list.Add((Document)MakeOne(typeId, typeConfig, seed, i));
                    var sec = new List<Type> { typeof(DocumentMeta), typeof(DocumentItem) };
                    if (n == 1) return (list[0], typeof(Document), sec);
                    sec.Insert(0, typeof(Document));
                    return (new BatchDocument { Items = list }, typeof(BatchDocument), sec);
                }
                case "telemetry":
                {
                    var list = new List<Telemetry>(n);
                    for (int i = 0; i < n; i++) list.Add((Telemetry)MakeOne(typeId, typeConfig, seed, i));
                    if (n == 1) return (list[0], typeof(Telemetry), new List<Type>());
                    return (new BatchTelemetry { Items = list }, typeof(BatchTelemetry),
                        new List<Type> { typeof(Telemetry) });
                }
                case "strings":
                {
                    var list = new List<Strings>(n);
                    for (int i = 0; i < n; i++) list.Add((Strings)MakeOne(typeId, typeConfig, seed, i));
                    if (n == 1) return (list[0], typeof(Strings), new List<Type>());
                    return (new BatchStrings { Items = list }, typeof(BatchStrings),
                        new List<Type> { typeof(Strings) });
                }
                default:
                    throw new ArgumentException($"unknown type_id: {typeId}");
            }
        }

        static Message MakeMessage(Rng r, JsonElement cfg)
        {
            var (lo, hi) = IRange(cfg);
            var (smin, smax) = SLen(cfg);
            return new Message
            {
                FBool = r.NextBool(),
                FInt32 = r.NextInt(lo, Math.Min(hi, int.MaxValue)),
                FInt64 = r.NextInt(lo, hi),
                FFloat64 = r.NextF64() * 1000.0,
                FString = r.Word(smin, smax),
                FBool2 = r.NextBool(),
                FInt322 = r.NextInt(lo, Math.Min(hi, int.MaxValue)),
                FString2 = r.Word(smin, smax),
            };
        }

        static Document MakeDocument(Rng r, JsonElement cfg)
        {
            int children = GetInt(cfg, "children", 8);
            var (smin, smax) = SLen(cfg);
            var items = new List<DocumentItem>(children);
            for (int i = 0; i < children; i++)
                items.Add(new DocumentItem
                {
                    Sku = r.Word(smin, smax),
                    Qty = r.NextInt(1, 100),
                    PriceMinor = r.NextInt(0, 100_000),
                });
            return new Document
            {
                Id = r.Word(8, 12),
                Status = r.NextInt(0, 5),
                Meta = new DocumentMeta { Region = r.Word(2, 4), Version = r.NextInt(1, 10) },
                Items = items,
            };
        }

        static Telemetry MakeTelemetry(Rng r, JsonElement cfg)
        {
            int points = GetInt(cfg, "points", 32);
            int tagCount = GetInt(cfg, "tag_count", 2);
            var (smin, smax) = SLen(cfg);
            var tags = new List<string>(tagCount);
            for (int i = 0; i < tagCount; i++) tags.Add(r.Word(smin, smax));
            var values = new List<double>(points);
            var numberType = GetString(cfg, "number_type", "float64");
            for (int i = 0; i < points; i++)
                values.Add(numberType == "int64" ? r.NextInt(0, 10_000) : r.NextF64() * 100.0);
            return new Telemetry
            {
                Source = r.Word(smin, smax),
                Ts = BaseTsMs + r.NextInt(0, 86_400_000),
                Tags = tags,
                Values = values,
            };
        }

        static Strings MakeStrings(Rng r, JsonElement cfg)
        {
            int count = GetInt(cfg, "count", 32);
            var (smin, smax) = SLen(cfg);
            double dup = GetDouble(cfg, "duplication", 0.1);
            var pool = new List<string>();
            var items = new List<string>(count);
            for (int i = 0; i < count; i++)
            {
                if (pool.Count > 0 && r.NextF64() < dup)
                    items.Add(pool[r.NextInt(0, pool.Count - 1)]);
                else
                {
                    var w = r.Word(smin, smax);
                    pool.Add(w);
                    items.Add(w);
                }
            }
            return new Strings { Items = items };
        }

        static Event MakeEvent(Rng r, JsonElement cfg)
        {
            int attrCount = GetInt(cfg, "attr_count", 4);
            var (smin, smax) = SLen(cfg);
            var attrs = new List<EventAttr>(attrCount);
            for (int i = 0; i < attrCount; i++)
                attrs.Add(new EventAttr { Key = r.Word(smin, smax), Value = r.Word(smin, smax) });
            return new Event
            {
                EventId = r.Word(8, 12),
                EventType = r.Word(smin, smax),
                OccurredAt = BaseTsMs + r.NextInt(0, 86_400_000),
                Producer = r.Word(smin, smax),
                Attrs = attrs,
            };
        }

        static (int min, int max) SLen(JsonElement cfg)
        {
            int min = 3, max = 16;
            if (cfg.ValueKind == JsonValueKind.Object && cfg.TryGetProperty("string_len", out var sl) &&
                sl.ValueKind == JsonValueKind.Object)
            {
                if (sl.TryGetProperty("min", out var a)) min = a.GetInt32();
                if (sl.TryGetProperty("max", out var b)) max = b.GetInt32();
            }
            return (min, max);
        }

        static (int min, int max) IRange(JsonElement cfg)
        {
            int min = 0, max = 1_000_000;
            if (cfg.ValueKind == JsonValueKind.Object && cfg.TryGetProperty("int_range", out var ir) &&
                ir.ValueKind == JsonValueKind.Object)
            {
                if (ir.TryGetProperty("min", out var a)) min = a.GetInt32();
                if (ir.TryGetProperty("max", out var b)) max = b.GetInt32();
            }
            return (min, max);
        }

        static int GetInt(JsonElement cfg, string name, int def) =>
            cfg.ValueKind == JsonValueKind.Object && cfg.TryGetProperty(name, out var p) && p.TryGetInt32(out var v)
                ? v : def;

        static double GetDouble(JsonElement cfg, string name, double def) =>
            cfg.ValueKind == JsonValueKind.Object && cfg.TryGetProperty(name, out var p) && p.TryGetDouble(out var v)
                ? v : def;

        static string GetString(JsonElement cfg, string name, string def) =>
            cfg.ValueKind == JsonValueKind.Object && cfg.TryGetProperty(name, out var p) && p.ValueKind == JsonValueKind.String
                ? p.GetString() ?? def : def;

        static ulong MixSeed(ulong seed, string typeId, int idx)
        {
            ulong h = seed;
            foreach (var ch in typeId) h = (h ^ ch) * 0x100000001B3UL;
            h ^= (ulong)idx * 0x9E3779B97F4A7C15UL;
            return h == 0 ? 1UL : h;
        }

        /// <summary>
        /// Deterministic xorshift64* (within-language only). Zero seed uses
        /// floor(2^64/φ)=0x9E3779B97F4A7C15 (golden ratio; nothing-up-my-sleeve).
        /// </summary>
        sealed class Rng
        {
            ulong _state;
            public Rng(ulong seed) { _state = seed == 0 ? 0x9E3779B97F4A7C15UL : seed; }
            public ulong NextU64()
            {
                var x = _state;
                x ^= x << 13; x ^= x >> 7; x ^= x << 17;
                _state = x;
                return x;
            }
            public int NextInt(int lo, int hi)
            {
                if (hi <= lo) return lo;
                return lo + (int)(NextU64() % (ulong)(hi - lo + 1));
            }
            public bool NextBool() => (NextU64() & 1) == 1;
            public double NextF64() => (NextU64() >> 11) * (1.0 / (1UL << 53));
            public string Word(int minL, int maxL)
            {
                int n = NextInt(minL, maxL);
                const string a = "abcdefghijklmnopqrstuvwxyz";
                var chars = new char[n];
                for (int i = 0; i < n; i++) chars[i] = a[(int)(NextU64() % 26)];
                return new string(chars);
            }
        }
    }
}
